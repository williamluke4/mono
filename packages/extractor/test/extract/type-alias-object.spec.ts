let docs: any

it.only('can be self recursive', () => {
  const docs = ctx.extract(`
    export type A = { a: A }
  `)
  expect(docs).toMatchSnapshot()
})

it('can be recursive within inline objects', () => {
  expect(
    ctx.extract(`
      export type A = { a: { aa: B }}
      export type B = { b: { bb: A }}
    `)
  ).toMatchSnapshot()
})

describe('can be a named export', () => {
  beforeAll(() => {
    docs = ctx.extract(`export type A = {}`)
  })
  it('shows up as a named export', () => {
    expect(docs).toMatchObject({
      modules: [{ namedExports: [{ name: 'A' }] }],
    })
  })
  it('its type is added to the type index', () => {
    expect(docs).toMatchObject({
      typeIndex: { '("/a").A': { kind: 'alias' } },
    })
  })
  it('refs the type index entry', () => {
    expect(docs).toMatchObject({
      modules: [
        {
          namedExports: [{ type: { kind: 'typeIndexRef', link: '("/a").A' } }],
        },
      ],
    })
  })
})

it('can be a return type ref', () => {
  expect(
    ctx.extract(`
      export function fa (): A               { return undefined as any }
             type A = { b: B      }
      export type B = { a: string }
    `)
  ).toMatchSnapshot()
})

describe('can be a parameter type ref', () => {
  beforeAll(() => {
    // docs = ctx.extract(`
    //   export function withA (b: B)          { return undefined as any }
    //          type B = { a: A }
    //   export type A = { b: B }
    // `)
  })
  it.todo('type gets added to type indx')
  it.todo('parameter refs the type index')
})

describe('lists', () => {
  it.todo('can be a list')
})
