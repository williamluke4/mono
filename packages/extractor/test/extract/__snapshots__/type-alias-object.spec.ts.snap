// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can be a return type ref 1`] = `
Object {
  "modules": Array [
    Object {
      "kind": "module",
      "location": Object {
        "absoluteFilePath": "/a.ts",
      },
      "mainExport": null,
      "name": "a",
      "namedExports": Array [
        Object {
          "isTerm": true,
          "isType": false,
          "kind": "export",
          "name": "fa",
          "type": Object {
            "hasProps": false,
            "isOverloaded": false,
            "kind": "callable",
            "props": Array [],
            "raw": Object {
              "nodeFullText": "export function fa(): A {
  return undefined as any;
}",
              "nodeText": "export function fa(): A {
  return undefined as any;
}",
              "typeText": "() => A",
            },
            "sigs": Array [
              Object {
                "kind": "sig",
                "params": Array [],
                "return": Object {
                  "kind": "typeIndexRef",
                  "link": "(\\"/a\\").A",
                },
              },
            ],
          },
        },
        Object {
          "isTerm": false,
          "isType": true,
          "kind": "export",
          "name": "B",
          "type": Object {
            "kind": "typeIndexRef",
            "link": "(\\"/a\\").B",
          },
        },
      ],
    },
  ],
  "typeIndex": Object {
    "(\\"/a\\").A": Object {
      "kind": "alias",
      "name": "A",
      "raw": Object {
        "nodeFullText": "type A = { b: B };",
        "nodeText": "type A = { b: B };",
        "typeText": "A",
      },
      "type": Object {
        "kind": "object",
        "props": Array [
          Object {
            "kind": "prop",
            "name": "b",
            "type": Object {
              "kind": "typeIndexRef",
              "link": "(\\"/a\\").B",
            },
          },
        ],
        "raw": Object {
          "nodeFullText": "type A = { b: B };",
          "nodeText": "type A = { b: B };",
          "typeText": "A",
        },
      },
    },
    "(\\"/a\\").B": Object {
      "kind": "alias",
      "name": "B",
      "raw": Object {
        "nodeFullText": "export type B = { a: string };",
        "nodeText": "export type B = { a: string };",
        "typeText": "B",
      },
      "type": Object {
        "kind": "object",
        "props": Array [
          Object {
            "kind": "prop",
            "name": "a",
            "type": Object {
              "kind": "primitive",
              "type": "string",
            },
          },
        ],
        "raw": Object {
          "nodeFullText": "export type B = { a: string };",
          "nodeText": "export type B = { a: string };",
          "typeText": "B",
        },
      },
    },
  },
}
`;

exports[`can be recursive within inline objects 1`] = `
Object {
  "modules": Array [
    Object {
      "kind": "module",
      "location": Object {
        "absoluteFilePath": "/a.ts",
      },
      "mainExport": null,
      "name": "a",
      "namedExports": Array [
        Object {
          "isTerm": false,
          "isType": true,
          "kind": "export",
          "name": "A",
          "type": Object {
            "kind": "typeIndexRef",
            "link": "(\\"/a\\").A",
          },
        },
        Object {
          "isTerm": false,
          "isType": true,
          "kind": "export",
          "name": "B",
          "type": Object {
            "kind": "typeIndexRef",
            "link": "(\\"/a\\").B",
          },
        },
      ],
    },
  ],
  "typeIndex": Object {
    "(\\"/a\\").A": Object {
      "kind": "alias",
      "name": "A",
      "raw": Object {
        "nodeFullText": "export type A = { a: { aa: B } };",
        "nodeText": "export type A = { a: { aa: B } };",
        "typeText": "A",
      },
      "type": Object {
        "kind": "object",
        "props": Array [
          Object {
            "kind": "prop",
            "name": "a",
            "type": Object {
              "kind": "object",
              "props": Array [
                Object {
                  "kind": "prop",
                  "name": "aa",
                  "type": Object {
                    "kind": "typeIndexRef",
                    "link": "(\\"/a\\").B",
                  },
                },
              ],
              "raw": Object {
                "nodeFullText": "{ aa: B }",
                "nodeText": "{ aa: B }",
                "typeText": "{ aa: B; }",
              },
            },
          },
        ],
        "raw": Object {
          "nodeFullText": "export type A = { a: { aa: B } };",
          "nodeText": "export type A = { a: { aa: B } };",
          "typeText": "A",
        },
      },
    },
    "(\\"/a\\").B": Object {
      "kind": "alias",
      "name": "B",
      "raw": Object {
        "nodeFullText": "export type B = { b: { bb: A } };",
        "nodeText": "export type B = { b: { bb: A } };",
        "typeText": "B",
      },
      "type": Object {
        "kind": "object",
        "props": Array [
          Object {
            "kind": "prop",
            "name": "b",
            "type": Object {
              "kind": "object",
              "props": Array [
                Object {
                  "kind": "prop",
                  "name": "bb",
                  "type": Object {
                    "kind": "typeIndexRef",
                    "link": "(\\"/a\\").A",
                  },
                },
              ],
              "raw": Object {
                "nodeFullText": "{ bb: A }",
                "nodeText": "{ bb: A }",
                "typeText": "{ bb: A; }",
              },
            },
          },
        ],
        "raw": Object {
          "nodeFullText": "export type B = { b: { bb: A } };",
          "nodeText": "export type B = { b: { bb: A } };",
          "typeText": "B",
        },
      },
    },
  },
}
`;

exports[`can be self recursive 1`] = `
Object {
  "modules": Array [
    Object {
      "isMain": true,
      "kind": "module",
      "location": Object {
        "absoluteFilePath": "/src/a.ts",
      },
      "mainExport": null,
      "name": "a",
      "namedExports": Array [
        Object {
          "isTerm": false,
          "isType": true,
          "kind": "export",
          "name": "A",
          "type": Object {
            "kind": "typeIndexRef",
            "link": "(a).A",
          },
        },
      ],
      "path": "/",
      "tsdoc": null,
    },
  ],
  "typeIndex": Object {
    "(a).A": Object {
      "kind": "alias",
      "name": "A",
      "raw": Object {
        "nodeFullText": "export type A = { a: A };",
        "nodeText": "export type A = { a: A };",
        "typeText": "A",
      },
      "tsdoc": null,
      "type": Object {
        "kind": "object",
        "props": Array [
          Object {
            "kind": "prop",
            "name": "a",
            "type": Object {
              "kind": "typeIndexRef",
              "link": "(a).A",
            },
          },
        ],
        "raw": Object {
          "nodeFullText": "export type A = { a: A };",
          "nodeText": "export type A = { a: A };",
          "typeText": "A",
        },
      },
    },
  },
}
`;
